api/controllers/export-data-2.js

```diff
@@ -0,0 +1,257 @@
+// PouchDB / promises. We should remove both copies of these functions and create
+// a shared lineage library that we can use here, as well as in the rest of api,
+// sentinel and webapp.
+//
+// lineage.js in sentinel is probably the most comprehensive lineage library
+// right now, though it would need to be ES5-a-fied if we wanted to use it on
+// the front end.
+//
+// https://github.com/medic/medic-webapp/issues/XXXX
+//
+const findContact = (contactRows, id) => {
+  return id && contactRows.find(contactRow => contactRow.id === id);
+};
+const hydrateDataRecords = result => {
+  const rows = result.rows;
+
+  const contactIds = [];
+  rows.forEach(row => {
+    let parent = row.doc.contact;
+    while(parent) {
+      if (parent._id) {
+        contactIds.push(parent._id);
+      }
+      parent = parent.parent;
+    }
+  });
+
+  if (!contactIds.length) {
+    return Promise.resolve(result);
+  }
+
+  return DB.allDocs({
+    keys: contactIds,
+    include_docs: true
+  }).then(results => {
+    rows.forEach(row => {
+      const contactId = row.doc.contact && row.doc.contact._id;
+      const contact = findContact(results.rows, contactId);
+      if (contact) {
+        row.doc.contact = contact.doc;
+      }
+      let parent = row.doc.contact;
+      while(parent) {
+        const parentId = parent.parent && parent.parent._id;
+        const found = findContact(results.rows, parentId);
+        if (found) {
+          parent.parent = found.doc;
+        }
+        parent = parent.parent;
+      }
+    });
+
+    return result;
+  });
+};
+
+
+/**
+ * Flattens a given object into an object where the keys are dot-notation
+ * paths to the flattened values:
+ * {
+ *   "foo": {
+ *     "bar": "smang"
+ *   }
+ * }
+ *
+ * becomes:
+ *
+ * {
+ *   "foo.bar": "smang"
+ * }
+ */
+// TODO: is there really not a utility that does this?
+const flatten = (fields, prepend=[]) =>
+  Object.keys(fields).reduce((acc, k) => {
+    const path = [...prepend, k];
+    if (typeof fields[k] === 'object' && fields[k]) {
+      return _.extend(acc, flatten(fields[k], path));
+    } else {
+      acc[path.join('.')] = fields[k] || '';
+      return acc;
+    }
+  }, {});
+
+// TODO: is there really not a utility that does this?
+const safeGet = (obj, fields) => {
+  let v = obj;
+  for (const f of fields) {
+    v = v[f];
+    if (!v) {
+      return undefined;
+    }
+  }
+  return v;
+};
+
+const CSV_MAPPER = {
+  reports: (filters) => {
+    const forms = (
+      filters &&
+      filters.forms &&
+      filters.forms.selected &&
+      _.pluck(filters.forms.selected, 'code'));
+
+    let p = Promise.resolve();
+    // If they have not selected any forms we need to get all available ones
+    if (!forms) {
+      p = p.then(() =>
+        DB.query('medic-client/reports_by_form', {
+          group: true
+        }).then(results => results.rows.map(r => r.key[0]))
+      );
+    } else {
+      p = p.then(() => forms);
+    }
```

Comment:
> To me, pulling functions like that out is a great way to break the code into manageable chunks rather than having a 60 line function

I think I have reversed on this over the years, and generally don't feel a drive to pull out code into functions unless they are going to be called from multiple places. I think because function names are never descriptive / clear enough unless the code is doing something generic, in which case it would probably be called from multiple places anyway. So because of this I find that I'm always delving into functions, and so moving code from being inline to being a function somewhere else in the file is just another link in my editor to follow and reverse out of to work out what code does.

Happy to do it though.

Reply:
We can talk about this some time. I much prefer it in general because as I said above "it breaks down the code into manageable chunks" which means I can completely understand what the 10 line function is doing then I don't have to read that code again in the near future. It also means unit testing is sane instead of having to mock out everything. And finally it reduces the scope of side effects to just the parameters you passed in.

Might be a better conversation to have over a beer!